src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -1603,10 +1603,99 @@ private bool canAllocateOnStack(CORINFO_CLASS_STRUCT_* cls)
             return result;
         }
 
+        /// <summary>
+        /// Managed implementation of CEEInfo::getClassAlignmentRequirementStatic
+        /// </summary>
+        public static int GetClassAlignmentRequirementStatic(DefType type)
+        {
+            int alignment = type.Context.Target.PointerSize;
+
+            if (type is MetadataType metadataType && metadataType.HasLayout())
+            {
+                if (metadataType.IsSequentialLayout || MarshalUtils.IsBlittableType(metadataType))
+                {
+                    alignment = metadataType.InstanceFieldAlignment.AsInt;
+                }
+            }
+
+            if (type.Context.Target.Architecture == TargetArchitecture.ARM &&
+                alignment < 8 && type.RequiresAlign8())
+            {
+                // If the structure contains 64-bit primitive fields and the platform requires 8-byte alignment for
+                // such fields then make sure we return at least 8-byte alignment. Note that it's technically possible
+                // to create unmanaged APIs that take unaligned structures containing such fields and this
+                // unconditional alignment bump would cause us to get the calling convention wrong on platforms such
+                // as ARM. If we see such cases in the future we'd need to add another control (such as an alignment
+                // property for the StructLayout attribute or a marshaling directive attribute for p/invoke arguments)
+                // that allows more precise control. For now we'll go with the likely scenario.
+                alignment = 8;
+            }
+
+            return alignment;
+        }
+
+        private Dictionary<DefType, bool> _doubleAlignHeuristicCache = new Dictionary<DefType, bool>();
+
+        //*******************************************************************************
+        //
+        // Heuristic to determine if we should have instances of this class 8 byte aligned
+        //
+        static bool ShouldAlign8(int dwR8Fields, int dwTotalFields)
+        {
+            return dwR8Fields*2>dwTotalFields && dwR8Fields>=2;
```

Comment:
I don't pretend I understand this condition but I assume it's a port from CoreCLR (with just one R8 field, we don't need the 8-alignment on ARM32?).

Reply:
Arm32 always aligns on 64bit boundaries in these circumstances. Its only X86 which has this odd heuristic driven behavior.